### PR TITLE
feat(graphql_analyze): useNamedOperation

### PR DESCRIPTION
**Pull request by vohoanglong0107**
## Summary

Disallow operations without name, since GraphQL clients usually use operations' name to cache.

## Test Plan

Added new tests to verify this rule's behavior

### CODE REVIEW

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                "Rename this "{state.operation_type}" to "{state.suggested_name}"."
+            }),
+        )
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<GraphqlRuleAction> {
+        let mut mutation = ctx.root().begin();
+        let node = ctx.query().clone();
+        let new_name = make::graphql_name_binding(make::ident(&state.suggested_name));
+        let new_node = node.clone().detach().with_name(Some(new_name));
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:85]: Do you need `detach` here?

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "Anonymous GraphQL operations are forbidden. Make sure to name your " {state.operation_type}"!"
+                },
+            )
+            .note(markup! {
+                "Rename this "{state.operation_type}" to "{state.suggested_name}"."
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:76]: Let's follow our rule pillars https://biomejs.dev/linter/#rule-pillars

This means that the .note must become a message that explains the error. The core action is the solution for fixing the error.

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    fn run(ctx: &RuleContext<Self>) -> Option<Self::State> {
+        let node = ctx.query();
+        let operation_type = node.ty().ok()?.text();
+        if node.name().is_some() {
+            None
+        } else {
+            Some(NoAnonymousOperationsState {
+                operation_type: operation_type.clone(),
+                suggested_name: get_suggested_name(node, operation_type),
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:59]: We should allocate strings only when needed. Here, we should do it in the `action` function

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+impl Rule for NoAnonymousOperations {
+    type Query = Ast<GraphqlOperationDefinition>;
+    type State = NoAnonymousOperationsState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Option<Self::State> {
+        let node = ctx.query();
+        let operation_type = node.ty().ok()?.text();
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:53]: Don't use `text` because it allocates a string, which isn't needed. Instead, use `syntax().text_trimmed()`

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            markup! {
+                "Rename this "{state.operation_type}" to "{state.suggested_name}"."
+            },
+            mutation,
+        ))
+    }
+}
+
+fn get_suggested_name(operation: &GraphqlOperationDefinition, operation_type: String) -> String {
+    let suggested_name = get_suggested_name_base_on_content(operation).unwrap_or(operation_type);
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:100]: I don't see the need for two separate functions, because they are used once. Let's put everything in one single function

**Review by ematipico — CHANGES_REQUESTED** [review]

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+use biome_console::markup;
+use biome_graphql_factory::make;
+use biome_graphql_syntax::{GraphqlOperationDefinition, GraphqlOperationType};
+use biome_rowan::{AstNode, BatchMutationExt};
+use biome_string_case::Case;
+
+use crate::GraphqlRuleAction;
+
+declare_lint_rule! {
+    /// Require specifying name for GraphQL operations.
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:14]:
```suggestion
    /// Enforce specifying the name of GraphQL operations.
```

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    ///
+    /// This is useful since most GraphQL client libraries are using the operation name for caching purposes.
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:16]:
```suggestion
    /// This is useful because most GraphQL client libraries use the operation name for caching purposes.
```

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    ///
+    /// ```graphql
+    /// query Human {
+    ///   name
+    /// }
+    /// ```
+    ///
+    pub NoAnonymousOperations {
+        version: "next",
+        name: "noAnonymousOperations",
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:36]: I read the documentation, and it seems the rule is more focused on the name of the operations. It doesn't say anything about operations that don't have a name.

Wouldn't it make more sense to call the rule `useNamedOperation`? Plus, let's use the singular

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    fn diagnostic(ctx: &RuleContext<Self>, operation_type: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "Anonymous GraphQL operations are forbidden. Make sure to name your " {operation_type.text()}"!"
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:69]:
```suggestion
                    "Anonymous GraphQL operations are forbidden. Make sure to name your " {operation_type.text()}"."
```

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                },
+            )
+            .note(markup! {
+                "Most GraphQL client libraries use operation name for caching purposes."
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:73]:
```suggestion
                "Most GraphQL client libraries use the operation name for caching purposes."
```

File: crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        }
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, operation_type: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.range(),
```
Comment (ematipico) [inline crates/biome_graphql_analyze/src/lint/nursery/no_anonymous_operations.rs:67]: I think highlighting the whole query can be too much. You used very small examples, but if you enable the rule in a big project, the whole thing is going to be red, which isn't ideal for the users. What do you think if we highlight the operation instead?

**Review by ematipico — APPROVED** [review]

**Comment by vohoanglong0107** [comment]
Oops, my bad, didn't mean to re-request a review from you there 😅
